monad-wireauth/src/filter.rs

```diff
@@ -0,0 +1,462 @@
+    time::Duration,
+};
+
+use lru::LruCache;
+use tracing::debug;
+
+use crate::state::State;
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum FilterAction {
+    Pass,
+    SendCookie,
+    Drop,
+}
+
+pub struct Filter {
+    counter: u64,
+    last_reset: Duration,
+    handshake_rate_limit: u64,
+    handshake_rate_reset_interval: Duration,
+    ip_request_history: LruCache<IpAddr, Duration>,
+    ip_rate_limit_window: Duration,
+    max_sessions_per_ip: usize,
+    low_watermark_sessions: usize,
+    high_watermark_sessions: usize,
+}
+
+impl Filter {
+    pub fn new(
+        handshake_rate_limit: u64,
+        handshake_rate_reset_interval: Duration,
+        ip_rate_limit_window: Duration,
+        ip_history_capacity: usize,
+        max_sessions_per_ip: usize,
+        low_watermark_sessions: usize,
+        high_watermark_sessions: usize,
+    ) -> Self {
+        Self {
+            counter: 0,
+            last_reset: Duration::ZERO,
+            handshake_rate_limit,
+            handshake_rate_reset_interval,
+            ip_request_history: LruCache::new(NonZeroUsize::new(ip_history_capacity).unwrap()),
+            ip_rate_limit_window,
+            max_sessions_per_ip,
+            low_watermark_sessions,
+            high_watermark_sessions,
+        }
+    }
+
+    pub fn tick(&mut self, duration_since_start: Duration) {
+        if duration_since_start.saturating_sub(self.last_reset)
+            >= self.handshake_rate_reset_interval
+        {
+            self.counter = 0;
+            self.last_reset = duration_since_start;
+        }
+    }
+
+    pub fn next_reset_time(&self) -> Duration {
+        self.last_reset + self.handshake_rate_reset_interval
+    }
+
+    pub fn apply(
+        &mut self,
+        state: &State,
+        remote_addr: SocketAddr,
+        duration_since_start: Duration,
+        cookie_valid: bool,
+    ) -> FilterAction {
+        self.counter += 1;
+
+        let total_sessions = state.total_sessions();
+        if total_sessions >= self.high_watermark_sessions {
+            debug!(
+                remote_addr = %remote_addr,
+                sessions = total_sessions,
+                high_watermark = self.high_watermark_sessions,
+                "high load - rejecting new handshake"
+            );
+            return FilterAction::Drop;
+        }
+
+        let under_load = self.counter >= self.handshake_rate_limit;
+
+        if under_load {
+            debug!(
+                remote_addr = %remote_addr,
+                counter = self.counter,
+                rate_limit = self.handshake_rate_limit,
+                "rate limit exceeded - dropping handshake"
+            );
+            return FilterAction::Drop;
+        }
+
+        if total_sessions < self.low_watermark_sessions {
+            return FilterAction::Pass;
+        }
+
+        if !cookie_valid {
+            return FilterAction::SendCookie;
+        }
+
+        let ip = remote_addr.ip();
```

Comment:
Since this potentionally handles IPv6 it could be wise to support ratelimiting based on subnets. For v6 devices comonly get a `/64` net and can easily rotate IPs.